primitives/state-machine/src/trie_backend_essence.rs

```diff
@@ -323,13 +327,18 @@ impl<S: TrieBackendStorage<H>, H: Hasher, C: AsLocalTrieCache<H>> TrieBackendEss
 	#[inline]
 	fn with_recorder_and_cache<R>(
 		&self,
-		_: Option<H::Out>,
+		storage_root: Option<H::Out>,
 		callback: impl FnOnce(
 			Option<&mut dyn TrieRecorder<H::Out>>,
 			Option<&mut dyn TrieCache<NodeCodec<H>>>,
 		) -> R,
 	) -> R {
-		callback(None, None)
+		let storage_root = storage_root.unwrap_or_else(|| self.root);
```

Comment:
I think we can now just merge both methods and put only the recorder stuff behind the feature flag. Should overall reduce the code.

Reply:
I experimented with this during the changes, but the code became much less readable with multiple feature checks (and additional `#[allow(unused)]` for the parameter), so I would like to leave them separated. In my opinion it is not worth the few lines we save here.

Reply:
Hmm... this doesn't seem that bad to me?

```rust
	#[inline]
	fn with_recorder_and_cache<R>(
		&self,
		storage_root: Option<H::Out>,
		callback: impl FnOnce(
			Option<&mut dyn TrieRecorder<H::Out>>,
			Option<&mut dyn TrieCache<NodeCodec<H>>>,
		) -> R,
	) -> R {
		let storage_root = storage_root.unwrap_or_else(|| self.root);
		let mut cache = self.trie_node_cache.as_ref().map(|c| c.as_trie_db_cache(storage_root));
		let cache = cache.as_mut().map(|c| c as _);

		#[cfg(not(feature = "std"))]
		{
			callback(None, cache)
		}

		#[cfg(feature = "std")]
		{
			let mut recorder = self.recorder.as_ref().map(|r| r.as_trie_recorder(storage_root));
			let recorder = match recorder.as_mut() {
				Some(recorder) => Some(recorder as &mut dyn TrieRecorder<H::Out>),
				None => None,
			};
			callback(recorder, cache)
		}
	}
```

Up to you.